walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -13,11 +13,83 @@
 // limitations under the License.
 
 package com.google.sps;
-
+import java.util.ArrayList;
+import java.util.Arrays;
 import java.util.Collection;
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.Set;
 
 public final class FindMeetingQuery {
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+    //throw new UnsupportedOperationException("TODO: Implement this method.");
+    // Create a bunch 
+    ArrayList<Event> eventsList = new ArrayList<>(events);
+    ArrayList<TimeRange> times = new ArrayList<TimeRange>();
+
+    HashSet<String> attendees = new HashSet<String>(request.getAttendees());
+    ArrayList<Event> sortedEvents = sortEvents(eventsList, attendees);
+    int requestDuration = (int) request.getDuration();
+    int eventStartTime = 0;
+    int index = 0;
+    TimeRange priorEventTime = null;
+    System.out.println("Adding events now!");
+    for (Event event : sortedEvents){
+        TimeRange currentEvent = event.getWhen();
+        int eventEndTime = currentEvent.end();
+
+        /* If this is is the first event or if this and the previous event
+            don't overlap, then add a new event to times */
+        if (priorEventTime == null || !priorEventTime.overlaps(currentEvent)){
+            TimeRange validEventTime =TimeRange.fromStartEnd(eventStartTime,currentEvent.start(), false);
+            if(validEventTime.duration() >= requestDuration){
+                times.add(validEventTime);
+            }
+            eventStartTime= eventEndTime;
+            priorEventTime = currentEvent;
+        } else if(priorEventTime != null && priorEventTime.contains(currentEvent)){
+            eventStartTime = priorEventTime.end(); 
+        } else{
+            eventStartTime = eventEndTime;
+            priorEventTime = currentEvent;
+        }
+    }
+    /* Add in event at end of day */
+    if(eventStartTime < 60*24 && 60*24 - requestDuration >= eventStartTime){
+        TimeRange validEventTime = TimeRange.fromStartEnd(eventStartTime,60*24, false);
+        times.add(validEventTime);
+    }
+    return times;
+  }
+
+    /**
+        Returns sorted list of events that attendees are attending.
+    */
+  public ArrayList<Event> sortEvents(ArrayList<Event> events, HashSet<String>attendees){
+    ArrayList<Event> sortedEvents = new ArrayList<Event>();
+    while(events.size()>0){
+        int minIndex = 0;
+        for(int i = 0; i < events.size(); i++){
+            if(events.get(i).getWhen().start() < events.get(minIndex).getWhen().start()){
+                minIndex=i;
+            }
+        }
```

Comment:
This seems pretty n^2 — maybe you can sort by start time, then filter?

Reply:
I tried Mergesort then filtering this time around.